### PR TITLE
feat(google-cloud-spanner): Implement "Inline Begin Transaction"

### PR DESCRIPTION
**Pull request by SandeepTuniki**
Implements "Inline Begin Transaction" optimisation for Ruby Spanner.
## Overview

Cloud Spanner allows users to query the data in several ways, including transactions. A transaction allows users to execute a series of statements as a single atomic unit, and allows them to rollback the changes in case of errors.

Most Cloud Spanner `*Request` APIs include a `TransactionSelector` field for queries. The `TransactionSelector` field can be used within queries to specify an existing transaction ID or start a new transaction.

Currently, we make an explicit `BeginTransaction` rpc to create a transaction in the backend, and get the transaction ID as the response. We pass the ID in the `TransactionSelector` field in the subsequent operations within the transaction.

This process has the following drawbacks:
* **Startup overhead** - The client sets up the environment  for transactions at the start of the client instantiation. This incurs an overhead in the instantiation time.

* **Prepared transactions** - As we cannot predict in advance how many transactions the user will need, we must prepare several transactions during client instantiation. This is done by maintaining two separate session pools in a specific ratio, one for read-only transactions and the other for read-write transactions. As the user accesses this pool of sessions, the client must rebalance the ratio of the two stacks.

* **Transaction latency** - The client makes an additional RPC to create a transaction. When the session pool runs out of available transactions, the client has to create more transactions. Both these factors increase the latency.

To address these drawbacks, we can defer the creation of transactions until the user executes the first query within the transaction. This is known as inlining the “begin transaction” call.

## Implementation details

Below are the the high-level list of changes done throughout the codebase:

### `Pool` class:
* Stop creating transactions in advance, and instead only maintain sessions in the pool.
* Remove the `write_ratio` field and the corresponding code that maintains transactions through write_ratio.
* Simplified maintenance:
  * Introduce new fields `sessions_available` and `sessions_in_use` to simplify the maintenance of session pool.
  * The new fields replace the previous fields `all_sessions`, `session_stack` and `transaction_stack`.

### `Transaction` class:
* Inline the transaction creation to the first operation within the transaction.
* Add thread safety mechanisms to ensure that the transaction creation isn't initiated multiple times by concurrent operations.
* Add some helper methods (ex: `existing_transaction?`, `safe_begin_transaction`, `safe_execute` etc).

### `Client` class:
* Refactor the `.transaction` method to handle inlining of transaction creation.
* Introduce helper methods (ex: `check_and_propagate_err()`) to reduce code complexity issues identified by Rubocop.

### `BatchUpdateResults` class:
* New class introduced to facilitate easier handling of `tx.batch_update()` API's response.

### `Project` class:
* Mark the `write_ratio` field as deprecated during client initialisation.

### `Results` class:
* Add a helper method `transaction` to facilitate extraction of a transaction from the results metadata.

### `Session` class:
* Add a helper method to create an empty transaction object (i.e; without gapic object), to facilitate ILB of a transaction.

### `Service` class:
* Refactor some code & move the processing logic (ex: batch update) to other classes.

### Unit tests
* Several unit tests are added to test the inlining of transactions.
* Many existing unit tests are modified to handle ILB.
* All the obsolete unit tests in the session pool which are related to `write_ratio` & `transaction_stack` are removed, and some are modified to fit the new mechanism of the session pool.

### Acceptance tets
* All existing acceptance should pass, since ILB is an internal optimisation.
* Some existing acceptance tests are modified for easier testing.
* Few tests are added to handle edge cases.

### CODE REVIEW

**Comment by SandeepTuniki** [comment]
Below are the benchmarking results after inlining the "begin transaction" rpc.

Scenario 1_ Client init + query

Scenario 2_ Query excluding client init

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb [hunk 1]
```diff
@@ -83,13 +84,21 @@ class Transaction
         def initialize
           @commit = Commit.new
           @seqno = 0
+
+          @mutex = Mutex.new
```
Comment (dazuma) [inline google-cloud-spanner/lib/google/cloud/spanner/transaction.rb:88]: We haven't done this very well in the existing code, but it's good practice to document the invariants enforced by mutexes because concurrency is hard on future maintainers.

**Review by dazuma — CHANGES_REQUESTED** [review]

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb (hunk 1 above)
Comment (SandeepTuniki) [reply google-cloud-spanner/lib/google/cloud/spanner/transaction.rb:90]: I documented the invariant at it's definition. Let me know if you prefer a more formal YARD style documentation (I avoided since it requires exposing the field through `attr_reader`).

File: google-cloud-spanner/lib/google/cloud/spanner/pool.rb [hunk 2]
```diff
@@ -182,22 +131,20 @@ def keepalive_or_release!
...
-            # Remove a random portion of the sessions and transactions
+            # Remove a random portion of the sessions
             to_release = to_keepalive.sample release_count
             to_keepalive -= to_release
 
             # Remove those to be released from circulation
-            @all_sessions -= to_release.map(&:session)
-            @session_stack -= to_release
-            @transaction_stack -= to_release
+            @sessions_available -= to_release
```
[the remark this answers is not in the document]
Comment (SandeepTuniki) [reply google-cloud-spanner/lib/google/cloud/spanner/pool.rb:147]: I updated all the references to use accessors (except a few instances where Rubocop was flagging redefining as local variables).

File: google-cloud-spanner/lib/google/cloud/spanner/pool.rb [hunk 3]
```diff
@@ -29,19 +29,15 @@ module Spanner
       # {Google::Cloud::Spanner::Session} instances.
       #
       class Pool
-        attr_accessor :all_sessions
-        attr_accessor :session_stack
-        attr_accessor :transaction_stack
+        attr_accessor :sessions_available
+        attr_accessor :sessions_in_use
```
[the remark this answers is not in the document]
Comment (SandeepTuniki) [reply google-cloud-spanner/lib/google/cloud/spanner/pool.rb:33]: They are used in unit tests to set the initial state:

```rb
pool.sessions_available = [session]
pool.sessions_in_use = []
```

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb [hunk 4]
```diff
@@ -83,13 +84,21 @@ class Transaction
...
+          @resource = ConditionVariable.new
+          @transaction_creation_in_progress = false
         end
 
         ##
         # Identifier of the transaction results were run in.
         # @return [String] The transaction id.
         def transaction_id
-          return nil if @grpc.nil?
+          return @grpc.id if existing_transaction?
```
[the remark this answers is not in the document]
Comment (SandeepTuniki) [reply google-cloud-spanner/lib/google/cloud/spanner/transaction.rb:97]: Missed this case, thanks for spotting it! I updated the code to handle the race condition, and added unit tests for this case.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb [hunk 5]
```diff
@@ -1788,7 +1789,7 @@ def commit commit_options: nil, request_options: nil,
         def transaction deadline: 120, commit_options: nil,
                         request_options: nil, call_options: nil
           ensure_service!
-          unless Thread.current[:transaction_id].nil?
+          unless Thread.current[:is_transaction_running].nil?
```
[the remark this answers is not in the document]
Comment (SandeepTuniki) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb:1792]: How about a key something like this? (I'm using the suffix `_KEY` to clarify that this isn't a boolean variable).

```rb
class Transaction
##
# @private
IS_TRANSACTION_RUNNING_KEY = "ruby_spanner_is_transaction_running".freeze

# ... some code in between ...

unless Thread.current[IS_TRANSACTION_RUNNING_KEY].nil?

# ...
```

[the remark this answers is not in the document]
Comment (dazuma) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Yeah something like that would be good.

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb (hunk 1 above)
Comment (dazuma) [reply google-cloud-spanner/lib/google/cloud/spanner/transaction.rb:90]: The documentation's location is fine; it's internal implementation documentation that doesn't need to be public. However, I was looking for something a bit more detailed on the actual invariant(s) being enforced, so that a future maintainer will know what kind of code needs to be protected by this mutex, and what they can assume when synchronizing on it. Especially because in this case there are multiple things being protected. My suggestion is something like this:

> This mutex protects two things: (1) the generation of sequence numbers, and (2) the creation of transactions. Specifically, `@seqno` is protected so it always reflects the last sequence number generated and provided to an operation in any thread. Any acquisition of a sequence number must be synchronized. Furthermore, `@grpc` is protected such that it either is `nil` if the transaction has not yet been created, or references the transaction resource if the transaction has been created. Any operation that could create a transaction must be synchronized, and any logic that depends on the state of transaction creation must also be synchronized.

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb [hunk 6]
```diff
@@ -1111,12 +1137,63 @@ def self.from_grpc grpc, session
...
+        # Create a new transaction in a thread-safe manner,
+        def safe_begin_transaction
+          @mutex.synchronize do
+            return if existing_transaction?
+            ensure_session!
+            @grpc = service.begin_transaction session.path
+          end
+        end
+
         # The TransactionSelector to be used for queries
```
Comment (dazuma) [inline google-cloud-spanner/lib/google/cloud/spanner/transaction.rb:1189]: It would probably be good to note in the documentation here that calls to this method must be synchronized by the mutex, since it includes logic that depends on the state of `existing_transaction?` which could otherwise be changed by another thread that owns the mutex. (Currently, it is properly called from within synchronize blocks because all calls are from blocks passed to `safe_execute`. We might even want to mention that that is the expected usage.)

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb [hunk 7]
```diff
@@ -1133,6 +1210,15 @@ def build_request_options options
           options
         end
 
+        ##
+        # @private Generates the next seqno in a thread-safe manner.
+        def next_seqno
```
Comment (dazuma) [inline google-cloud-spanner/lib/google/cloud/spanner/transaction.rb:1215]: Maybe consider naming this `safe_next_seqno` to emphasize that we're synchronizing within this method

**Review by dazuma — APPROVED** [review]
I think the logic looks good now. I have a few more recommendations on internal documentation, just to help with maintainability of the concurrency and synchronization.

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb (hunk 1 above)
Comment (SandeepTuniki) [reply google-cloud-spanner/lib/google/cloud/spanner/transaction.rb:106]: Thanks for the writeup :) I updated the file with the documentation.

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb (hunk 6 above)
Comment (SandeepTuniki) [reply google-cloud-spanner/lib/google/cloud/spanner/transaction.rb]: I updated the documentation with the suggested info.

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb (hunk 7 above)
Comment (SandeepTuniki) [reply google-cloud-spanner/lib/google/cloud/spanner/transaction.rb]: Done.

**Review by NivedhaSenthil — APPROVED** [review]
